spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java

```diff
@@ -99,47 +60,42 @@ public ClientHttpResponse intercept(final HttpRequest request, final byte[] body
+		BackOffPolicy backOffPolicy = lbRetryFactory.createBackOffPolicy(serviceName);
 		template.setBackOffPolicy(backOffPolicy == null ? new NoBackOffPolicy() : backOffPolicy);
 		template.setThrowLastExceptionOnExhausted(true);
-		RetryListener[] retryListeners = this.retryListenerFactory.createRetryListeners(serviceName);
+		RetryListener[] retryListeners = lbRetryFactory.createRetryListeners(serviceName);
                if (retryListeners != null && retryListeners.length != 0) {
                    template.setListeners(retryListeners);
                }
 		template.setRetryPolicy(
 				!lbProperties.isEnabled() || retryPolicy == null ? new NeverRetryPolicy()
 						: new InterceptorRetryPolicy(request, retryPolicy, loadBalancer,
 						serviceName));
-		return template
-				.execute(new RetryCallback<ClientHttpResponse, IOException>() {
-					@Override
-					public ClientHttpResponse doWithRetry(RetryContext context)
-							throws IOException {
-						ServiceInstance serviceInstance = null;
-						if (context instanceof LoadBalancedRetryContext) {
-							LoadBalancedRetryContext lbContext = (LoadBalancedRetryContext) context;
-							serviceInstance = lbContext.getServiceInstance();
-						}
-						if (serviceInstance == null) {
-							serviceInstance = loadBalancer.choose(serviceName);
-						}
-						ClientHttpResponse response = RetryLoadBalancerInterceptor.this.loadBalancer.execute(
-								serviceName, serviceInstance,
-								requestFactory.createRequest(request, body, execution));
-						int statusCode = response.getRawStatusCode();
-						if (retryPolicy != null && retryPolicy.retryableStatusCode(statusCode)) {
-							throw new ClientHttpResponseStatusCodeException(serviceName, response);
-						}
-						return response;
-					}
-				}, new RibbonRecoveryCallback<ClientHttpResponse, ClientHttpResponse>() {
-					@Override
-					protected ClientHttpResponse createResponse(ClientHttpResponse response, URI uri) {
-						return response;
-					}
-				});
+		return template.execute(context -> {
+			ServiceInstance serviceInstance = null;
+			if (context instanceof LoadBalancedRetryContext) {
+				LoadBalancedRetryContext lbContext = (LoadBalancedRetryContext) context;
+				serviceInstance = lbContext.getServiceInstance();
+			}
+			if (serviceInstance == null) {
+				serviceInstance = loadBalancer.choose(serviceName);
+			}
+			ClientHttpResponse response = RetryLoadBalancerInterceptor.this.loadBalancer.execute(
+					serviceName, serviceInstance,
+					requestFactory.createRequest(request, body, execution));
+			int statusCode = response.getRawStatusCode();
+			if (retryPolicy != null && retryPolicy.retryableStatusCode(statusCode)) {
+				throw new ClientHttpResponseStatusCodeException(serviceName, response);
+			}
+			return response;
+		}, new LoadBalancedRecoveryCallback<ClientHttpResponse, ClientHttpResponse>() {
+			@Override
+			protected ClientHttpResponse createResponse(ClientHttpResponse response, URI uri) {
```

Comment:
Looks like this could be pushed up as the default impl of the abstract method too. Simplify even more.

Reply:
I dont think so.  This is a special case where both parameters are the same type.  In every other case (in Netflix and Feign) the two type parameters are different.

Reply:
Maybe a comment then

Reply:
done